packages/@aws-cdk/aws-codebuild-artifacts/lib/artifacts.ts

```diff
@@ -0,0 +1,84 @@
+import codebuild = require("@aws-cdk/aws-codebuild");
+import s3 = require("@aws-cdk/aws-s3");
+
+/**
+ * CodePipeline Artifact definition for a CodeBuild Project.
+ * *Note*: this type cannot be used as a secondary artifact,
+ * and because of that, you're not allowed to specify an identifier for it.
+ */
+export class CodePipelineBuildArtifacts extends codebuild.BuildArtifacts {
+  protected readonly type = codebuild.CODEPIPELINE_SOURCE_ARTIFACT_TYPE;
+
+  constructor() {
+    super({});
+  }
+}
+
+/**
+ * Construction properties for {@link S3BucketBuildArtifacts}.
+ */
+export interface S3BucketBuildArtifactsProps extends codebuild.BuildArtifactsProps {
+  /**
+   * The name of the output bucket.
+   */
+  readonly bucket: s3.IBucket;
+
+  /**
+   * The path inside of the bucket for the build output .zip file or folder.
+   * If a value is not specified, then build output will be stored at the root of the
+   * bucket (or under the <build-id> directory if `includeBuildId` is set to true).
+   */
+  readonly path?: string;
+
+  /**
+   * The name of the build output ZIP file or folder inside the bucket.
+   *
+   * The full S3 object key will be "<path>/<build-id>/<name>" or
+   * "<path>/<name>" depending on whether `includeBuildId` is set to true.
+   */
+  readonly name: string;
+
+  /**
+   * Indicates if the build ID should be included in the path. If this is set to true,
+   * then the build artifact will be stored in "<path>/<build-id>/<name>".
+   *
+   * @default true
+   */
+  readonly includeBuildId?: boolean;
+
+  /**
+   * If this is true, all build output will be packaged into a single .zip file.
+   * Otherwise, all files will be uploaded to <path>/<name>
+   *
+   * @default true - files will be archived
+   */
+  readonly packageZip?: boolean;
+}
+
+/**
+ * S3 Artifact definition for a CodeBuild Project.
+ */
+export class S3BucketBuildArtifacts extends codebuild.BuildArtifacts {
+  protected readonly type = 'S3';
+
+  constructor(private readonly props: S3BucketBuildArtifactsProps) {
+    super(props);
+  }
+
+  /**
+   * @internal
+   */
+  public _bind(project: codebuild.Project) {
```

Comment:
Since we're touching this anyway... the `bind()` pattern changed a little:

* We implement an interface `IBuildArtifact`.
* `bind` has signature `bind(scope: Construct, project: codebuild.IProject): BuildArtifactConfig`.
  * Project interface, not class.
  * `toArtifactsProperty()` can be removed, it becomes the output of `bind()`.
* Doesn't have to be `@internal` anymore.

Reply:
Done, for both `Artifacts` and `Source`.